changelog.md

```diff
@@ -1,3 +1,68 @@
+## Code Dx Jenkins Plugin 4.0.0 `Released 5/10/2023`
+
+### Changes
+- The "API Key" field now uses a `credential` instead of a `string` value
+- The "API Key" field ID (`key`) has been renamed to `keyCredentialId`
+- When configuring Freestyle projects, "Configure" permission on the project is required to make changes
+- When using the Pipeline Syntax Generator tool, the global "Administer" permission is required to validate the form and list projects
+  - It can still be used without these permissions, but the "Code Dx Project" field will need to be filled manually since the drop-down will remain empty
+
+### Notice
+
+This release was performed in cooperation with the Jenkins team and is a response to several reported vulnerabilities:
+
+- [CSRF vulnerability and missing permission checks in Code Dx Plugin](https://www.jenkins.io/security/advisory/2023-05-16/#SECURITY-3118)
+- [Missing permission checks in Code Dx Plugin](https://www.jenkins.io/security/advisory/2023-05-16/#SECURITY-3145)
+- [API keys stored and displayed in plain text by Code Dx Plugin](https://www.jenkins.io/security/advisory/2023-05-16/#SECURITY-3146)
+
+It is a breaking change and will require reconfiguring Pipeling and Freestyle Projects.
+
+#### Freestyle Projects
+
+API Keys must be stored in Jenkins Credentials as "Secret text". Each project must be reconfigured in order to assign an API Key Credential.
+
+Plugin version 4.0.0 has several known bugs when configuring via a form in a Freestyle Project or the Pipeline Syntax Generator. In particular, when reconfiguring a Freestyle Project, **_the selected project may be reset and should be verified before saving your updated configuration._**
+
+In future releases which fix these bugs, the selected project should still be confirmed before saving your changes. These concerns only apply when upgrading from plugin version 3.x or earlier to version 4.0.0.
```

Comment:
I know this is for the 4.0.0 release so maybe this is premature, but I wonder if "... to version 4.0.0" should be changed to "... to version 4.0.0 and greater" (or similar wording)?

Reply:
This should only be an issue when upgrading from 3.x or earlier to 4.0.0 specifically, since the 4.0.0 release has the change which causes the form to become invalid (and potentially reset the old value)

Once you've upgraded to 4.0.0 and made your changes, upgrades to later 4.x releases shouldn't have the same problem since they'll be using the same form config (which won't lead to errors or config field reset)

Reply:
My wording isn't great. For example, in the future, someone upgrades from 3.x or earlier to 4.0.1 (or 5.0.0).

Reply:
Gotcha', I'll update to say 4.x and we can revise this once the bugfix is released